File: src-tauri/src/commands/usage.rs
use crate::db::{
    self, AppState, ColumnCost, DailyCost, ModelUsageSummary, TaskCost, UsageRecord, UsageSummary,
};
use crate::error::AppError;
use tauri::State;

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn record_usage(
    state: State<AppState>,
    workspace_id: String,
    task_id: Option<String>,
    session_id: Option<String>,
    provider: String,
    model: String,
    input_tokens: i64,
    output_tokens: i64,
    cost_usd: f64,
    column_name: Option<String>,
    duration_seconds: Option<i64>,
) -> Result<UsageRecord, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::insert_usage_record(
        &conn,
        &workspace_id,
        task_id.as_deref(),
        session_id.as_deref(),
        &provider,
        &model,
        input_tokens,
        output_tokens,
        cost_usd,
        column_name.as_deref(),
        duration_seconds.unwrap_or(0),
    )
    .map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_usage(
    state: State<AppState>,
    workspace_id: String,
    limit: Option<i64>,
) -> Result<Vec<UsageRecord>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::list_usage_records(&conn, &workspace_id, limit).map_err(AppError::from)
}

#[tauri::command]
pub fn get_task_usage(
    state: State<AppState>,
    task_id: String,
) -> Result<Vec<UsageRecord>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::list_task_usage(&conn, &task_id).map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_usage_summary(
    state: State<AppState>,
    workspace_id: String,
) -> Result<UsageSummary, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_workspace_usage_summary(&conn, &workspace_id).map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_model_usage_between(
    state: State<AppState>,
    workspace_id: String,
    start_at: String,
    end_at: String,
) -> Result<Vec<ModelUsageSummary>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_workspace_model_usage_between(&conn, &workspace_id, &start_at, &end_at)
        .map_err(AppError::from)
}

#[tauri::command]
pub fn get_task_usage_summary(
    state: State<AppState>,
    task_id: String,
) -> Result<UsageSummary, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_task_usage_summary(&conn, &task_id).map_err(AppError::from)
}

#[tauri::command]
pub fn clear_workspace_usage(state: State<AppState>, workspace_id: String) -> Result<(), AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::delete_workspace_usage(&conn, &workspace_id).map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_daily_costs(
    state: State<AppState>,
    workspace_id: String,
    days: Option<i64>,
) -> Result<Vec<DailyCost>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_workspace_daily_costs(&conn, &workspace_id, days.unwrap_or(30)).map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_column_costs(
    state: State<AppState>,
    workspace_id: String,
) -> Result<Vec<ColumnCost>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_workspace_column_costs(&conn, &workspace_id).map_err(AppError::from)
}

#[tauri::command]
pub fn get_workspace_task_costs(
    state: State<AppState>,
    workspace_id: String,
    limit: Option<i64>,
) -> Result<Vec<TaskCost>, AppError> {
    let conn = state
        .db
        .lock()
        .map_err(|e| AppError::DatabaseError(e.to_string()))?;
    db::get_workspace_task_costs(&conn, &workspace_id, limit.unwrap_or(10)).map_err(AppError::from)
}

File: src/lib/ipc/usage.ts
import { invoke } from './invoke'

// ─── Usage tracking commands ─────────────────────────────────────────────────

export type UsageRecord = {
  id: string
  workspaceId: string
  taskId: string | null
  sessionId: string | null
  provider: string
  model: string
  inputTokens: number
  outputTokens: number
  costUsd: number
  columnName: string | null
  durationSeconds: number
  createdAt: string
}

export type UsageSummary = {
  totalInputTokens: number
  totalOutputTokens: number
  totalCostUsd: number
  recordCount: number
}

export type WorkspaceCostSummary = {
  workspaceId: string
  workspaceName: string
  totalCostUsd: number
  totalInputTokens: number
  totalOutputTokens: number
  recordCount: number
}

export type ColumnCostSummary = {
  workspaceId: string
  workspaceName: string
  columnName: string
  totalCostUsd: number
  totalInputTokens: number
  totalOutputTokens: number
  recordCount: number
}

export type TaskCostSummary = {
  taskId: string | null
  taskTitle: string
  workspaceId: string
  workspaceName: string
  columnName: string
  totalCostUsd: number
  totalInputTokens: number
  totalOutputTokens: number
  recordCount: number
}

export type DailyCostSummary = {
  date: string
  totalCostUsd: number
  totalInputTokens: number
  totalOutputTokens: number
  recordCount: number
}

export type CostDashboard = {
  total: UsageSummary
  workspaces: WorkspaceCostSummary[]
  columns: ColumnCostSummary[]
  topTasks: TaskCostSummary[]
  daily: DailyCostSummary[]
}

export async function recordUsage(
  workspaceId: string,
  provider: string,
  model: string,
  inputTokens: number,
  outputTokens: number,
  costUsd: number,
  taskId?: string,
  sessionId?: string,
  columnName?: string,
  durationSeconds?: number,
): Promise<UsageRecord> {
  return invoke<UsageRecord>('record_usage', {
    workspaceId,
    taskId,
    sessionId,
    provider,
    model,
    inputTokens,
    outputTokens,
    costUsd,
    columnName,
    durationSeconds,
  })
}

export async function getWorkspaceUsage(
  workspaceId: string,
  limit?: number,
): Promise<UsageRecord[]> {
  return invoke<UsageRecord[]>('get_workspace_usage', { workspaceId, limit })
}

export async function getTaskUsage(taskId: string): Promise<UsageRecord[]> {
  return invoke<UsageRecord[]>('get_task_usage', { taskId })
}

export async function getWorkspaceUsageSummary(
  workspaceId: string,
): Promise<UsageSummary> {
  return invoke<UsageSummary>('get_workspace_usage_summary', { workspaceId })
}

export async function getTaskUsageSummary(taskId: string): Promise<UsageSummary> {
  return invoke<UsageSummary>('get_task_usage_summary', { taskId })
}

export async function clearWorkspaceUsage(workspaceId: string): Promise<void> {
  return invoke('clear_workspace_usage', { workspaceId })
}

export async function getCostDashboard(): Promise<CostDashboard> {
  return invoke<CostDashboard>('get_cost_dashboard')
}
